src/btree/bt_read.c

```diff
@@ -65,261 +65,79 @@ __row_instantiate(
-            break;
-        default:
-            WT_ERR(__wt_illegal_value(session, page->type));
-        }
-
-        /* Append the latest update to the list. */
-        if (first_upd == NULL)
-            first_upd = last_upd = upd;
-        else {
-            last_upd->next = upd;
-            last_upd = upd;
-        }
-        upd = NULL;
-    }
-    __wt_readunlock(session, &cache->las_sweepwalk_lock);
-    locked = false;
-    WT_ERR_NOTFOUND_OK(ret);
-
-    /* Insert the last set of updates, if any. */
-    if (first_upd != NULL) {
-        WT_ASSERT(session, __wt_count_birthmarks(first_upd) <= 1);
-        switch (page->type) {
-        case WT_PAGE_COL_FIX:
-        case WT_PAGE_COL_VAR:
-            WT_ERR(__col_instantiate(session, current_recno, ref, &cbt, first_upd));
-            first_upd = NULL;
-            break;
-        case WT_PAGE_ROW_LEAF:
-            WT_ERR(__row_instantiate(session, current_key, ref, &cbt, first_upd));
-            first_upd = NULL;
+            WT_ERR(__row_instantiate(session, &birthmarkp->key, ref, &cbt, upd));
+            upd = NULL;
             break;
-        default:
-            WT_ERR(__wt_illegal_value(session, page->type));
         }
     }
 
-    /* Discard the cursor. */
-    WT_ERR(__wt_las_cursor_close(session, &cursor, session_flags));
-
-    if (total_incr != 0) {
-        __wt_cache_page_inmem_incr(session, page, total_incr);
-
-        /*
-         * If the updates in lookaside are newer than the versions on
-         * the page, it must be included in the next checkpoint.
-         *
-         * Otherwise, the page image contained the newest versions of
-         * data so the updates are all older and we could consider
-         * marking it clean (i.e., the next checkpoint can use the
-         * version already on disk).
-         *
-         * This needs care because (a) it creates pages with history
-         * that can't be evicted until they are marked dirty again, and
-         * (b) checkpoints may need to visit these pages to resolve
-         * changes evicted while a checkpoint is running.
-         */
-        page->modify->first_dirty_txn = WT_TXN_FIRST;
-
-        FLD_SET(page->modify->restore_state, WT_PAGE_RS_LOOKASIDE);
-
-        if (ref->page_las->skew_newest && !ref->page_las->has_prepares &&
-          !S2C(session)->txn_global.has_stable_timestamp &&
-          __wt_txn_visible_all(
-              session, ref->page_las->unstable_txn, ref->page_las->unstable_durable_timestamp)) {
-            page->modify->rec_max_txn = ref->page_las->max_txn;
-            page->modify->rec_max_timestamp = ref->page_las->max_timestamp;
-            __wt_page_modify_clear(session, page);
-        }
+    /* We do not need the birthmark information in the lookaside structure anymore. */
+    for (i = 0; i < ref->page_las->birthmarks_cnt; i++) {
+        birthmarkp = ref->page_las->birthmarks + i;
+        __wt_buf_free(session, &birthmarkp->key);
```

Comment:
In case of the column store as well we are allocating memory for a key:
https://github.com/wiredtiger/wiredtiger/blob/wt-5058-no-las/src/cache/cache_las.c#L722

I am only freeing these saved birthmarks if the process of instantiating all birthmarks succeed. If we free them as we instantitate per key, and fail later, we will lose the birthmark information in case the page is tried to be read to memory again later.

Reply:
Got it, sounds right -- thanks!